collectors/state_collector.go

```diff
@@ -0,0 +1,104 @@
+package collectors
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/lightninglabs/lndclient"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type StateCollector struct {
+	lnd *lndclient.LndServices
+
+	// Use one gauge to track the starting time of LND.
+	timeToStartDesc *prometheus.Desc
+
+	// startTime records a best-effort timestamp of when LND was started.
+	startTime time.Time
+
+	// endTime records when LND makes a transition from RPC_ACTIVE to
+	// SERVER_ACTIVE.
+	endTime time.Time
+
+	// mutex is a lock for preventing concurrent writes to startTime or
+	// endTime.
+	mutex sync.RWMutex
+
+	// errChan is a channel that we send any errors that we encounter into.
+	// This channel should be buffered so that it does not block sends.
+	errChan chan<- error
+}
+
+// NewStateCollector returns a new instance of the StateCollector.
+func NewStateCollector(lnd *lndclient.LndServices,
+	errChan chan<- error) *StateCollector {
+
+	sc := &StateCollector{
+		lnd: lnd,
+		timeToStartDesc: prometheus.NewDesc(
+			"lnd_time_to_start_secs",
+			"time to start in seconds",
+			nil, nil,
+		),
+		errChan: errChan,
+	}
+
+	go sc.monitorStateChanges()
+	return sc
+}
+
+// monitorStateChanges checks the state every second to catch fast transitions.
+func (s *StateCollector) monitorStateChanges() {
+	for {
+		state, err := s.lnd.State.GetState(context.Background())
```

Comment:
Have we tested this in the wild? It seems to make an assumption that the `lndmon` sidecar container is _always_ started either before lnd, or before we do the unlock itself.

Reply:
Indeed that assumption is made, however it's a best effort metric.

If lndmon starts after LND has already reached SERVER_ACTIVE, no startup time metric will be set.

After all it's not possible for us to know the startup time if we are external to LND.

The perfect approach would probably be to instrument LND itself, but short of that this is a pretty good stopgap solution for some basic tests that implicitly guarantee lndmon and lnd start together.